File: TheepotMobile/TheepotMobile/Services/AgendaService.swift
import Foundation

/// Spiegelt app/agenda/page.tsx (zonder ICS-import/-abonneren, dat blijft
/// een web-only functie). Zichtbaarheid van kalenders volgt dezelfde regels
/// als de webapp: eigen persoonlijke kalender + (voor bevoorrechte rollen)
/// alle personeelskalenders + algemene kalenders (rechtstreeks zichtbaar voor
/// bevoorrechte rollen, anders alleen via `agenda_gedeeld`).
enum AgendaService {
    static func kalenders(profielId: String, magAlleKalendersZien: Bool, magAllePersoonlijkZien: Bool) async throws -> [AgendaKalender] {
        var persoonlijk: [AgendaKalender] = try await SupabaseManager.client
            .from("agenda_kalenders")
            .select("*")
            .eq("type", value: "persoonlijk")
            .eq("eigenaar_id", value: profielId)
            .execute()
            .value

        if persoonlijk.isEmpty {
            // Nieuwe gebruiker zonder persoonlijke kalender: eenmalig aanmaken,
            // zelfde als bij de eerste keer dat de webagenda ooit werd geopend.
            persoonlijk = [try await maakPersoonlijkeKalender(profielId: profielId)]
        }

        var allePersoonlijk = persoonlijk
        if magAllePersoonlijkZien {
            allePersoonlijk = try await SupabaseManager.client
                .from("agenda_kalenders")
                .select("*")
                .eq("type", value: "persoonlijk")
                .execute()
                .value
        }

        var algemeen: [AgendaKalender] = []
        if magAlleKalendersZien {
            algemeen = try await SupabaseManager.client
                .from("agenda_kalenders")
                .select("*")
                .eq("type", value: "algemeen")
                .execute()
                .value
        } else {
            struct GedeeldRij: Decodable { let kalender_id: String }
            let gedeeld: [GedeeldRij] = try await SupabaseManager.client
                .from("agenda_gedeeld")
                .select("kalender_id")
                .eq("profiel_id", value: profielId)
                .execute()
                .value
            if !gedeeld.isEmpty {
                algemeen = try await SupabaseManager.client
                    .from("agenda_kalenders")
                    .select("*")
                    .in("id", values: gedeeld.map(\.kalender_id))
                    .execute()
                    .value
            }
        }

        return allePersoonlijk + algemeen
    }

    private static func maakPersoonlijkeKalender(profielId: String) async throws -> AgendaKalender {
        struct Insert: Encodable { let naam: String; let type: String; let eigenaar_id: String; let kleur: String }
        return try await SupabaseManager.client
            .from("agenda_kalenders")
            .insert(Insert(naam: "Mijn agenda", type: "persoonlijk", eigenaar_id: profielId, kleur: "#4F46E5"))
            .select()
            .single()
            .execute()
            .value
    }

    static func afspraken(kalenderIds: [String]) async throws -> [AgendaAfspraak] {
        guard !kalenderIds.isEmpty else { return [] }
        return try await SupabaseManager.client
            .from("agenda_afspraken")
            .select("*")
            .in("kalender_id", values: kalenderIds)
            .order("start_tijd")
            .execute()
            .value
    }

    static func maakAfspraak(kalenderId: String, titel: String, beschrijving: String?, startTijd: String, eindTijd: String, heleDag: Bool, herinneringMinuten: Int?, aangemaaktDoor: String) async throws {
        struct Insert: Encodable {
            let kalender_id: String, titel: String, beschrijving: String?
            let start_tijd: String, eind_tijd: String, hele_dag: Bool
            let herinnering_minuten: Int?, aangemaakt_door: String
        }
        try await SupabaseManager.client
            .from("agenda_afspraken")
            .insert(Insert(kalender_id: kalenderId, titel: titel, beschrijving: beschrijving, start_tijd: startTijd, eind_tijd: eindTijd, hele_dag: heleDag, herinnering_minuten: herinneringMinuten, aangemaakt_door: aangemaaktDoor))
            .execute()
    }

    static func werkAfspraakBij(id: String, kalenderId: String, titel: String, beschrijving: String?, startTijd: String, eindTijd: String, heleDag: Bool, herinneringMinuten: Int?) async throws {
        struct Update: Encodable {
            let kalender_id: String, titel: String, beschrijving: String?
            let start_tijd: String, eind_tijd: String, hele_dag: Bool
            let herinnering_minuten: Int?
        }
        try await SupabaseManager.client
            .from("agenda_afspraken")
            .update(Update(kalender_id: kalenderId, titel: titel, beschrijving: beschrijving, start_tijd: startTijd, eind_tijd: eindTijd, hele_dag: heleDag, herinnering_minuten: herinneringMinuten))
            .eq("id", value: id)
            .execute()
    }

    static func verwijderAfspraak(id: String) async throws {
        try await SupabaseManager.client
            .from("agenda_afspraken")
            .delete()
            .eq("id", value: id)
            .execute()
    }
}

File: TheepotAndroid/app/src/main/kotlin/nl/bsodetheepot/mobile/data/services/AgendaService.kt
package nl.bsodetheepot.mobile.data.services

import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.serialization.Serializable
import nl.bsodetheepot.mobile.data.models.AgendaAfspraak
import nl.bsodetheepot.mobile.data.models.AgendaKalender

/**
 * Spiegelt app/agenda/page.tsx (zonder ICS-import/-abonneren, dat blijft een
 * web-only functie). Zichtbaarheid van kalenders volgt dezelfde regels als de
 * webapp: eigen persoonlijke kalender + (voor bevoorrechte rollen) alle
 * personeelskalenders + algemene kalenders (rechtstreeks zichtbaar voor
 * bevoorrechte rollen, anders alleen via `agenda_gedeeld`).
 */
object AgendaService {
    suspend fun kalenders(profielId: String, magAlleKalendersZien: Boolean, magAllePersoonlijkZien: Boolean): List<AgendaKalender> {
        var persoonlijk = SupabaseManager.client.postgrest["agenda_kalenders"]
            .select { filter { eq("type", "persoonlijk"); eq("eigenaar_id", profielId) } }
            .decodeList<AgendaKalender>()

        if (persoonlijk.isEmpty()) {
            // Nieuwe gebruiker zonder persoonlijke kalender: eenmalig aanmaken,
            // zelfde als bij de eerste keer dat de webagenda ooit werd geopend.
            persoonlijk = listOf(maakPersoonlijkeKalender(profielId))
        }

        var allePersoonlijk = persoonlijk
        if (magAllePersoonlijkZien) {
            allePersoonlijk = SupabaseManager.client.postgrest["agenda_kalenders"]
                .select { filter { eq("type", "persoonlijk") } }
                .decodeList<AgendaKalender>()
        }

        val algemeen = if (magAlleKalendersZien) {
            SupabaseManager.client.postgrest["agenda_kalenders"]
                .select { filter { eq("type", "algemeen") } }
                .decodeList<AgendaKalender>()
        } else {
            @Serializable
            data class GedeeldRij(val kalender_id: String)

            val gedeeld = SupabaseManager.client.postgrest["agenda_gedeeld"]
                .select(Columns.raw("kalender_id")) { filter { eq("profiel_id", profielId) } }
                .decodeList<GedeeldRij>()

            if (gedeeld.isEmpty()) {
                emptyList()
            } else {
                SupabaseManager.client.postgrest["agenda_kalenders"]
                    .select { filter { isIn("id", gedeeld.map { it.kalender_id }) } }
                    .decodeList<AgendaKalender>()
            }
        }

        return allePersoonlijk + algemeen
    }

    @Serializable
    private data class KalenderInsert(val naam: String, val type: String, val eigenaar_id: String, val kleur: String)

    private suspend fun maakPersoonlijkeKalender(profielId: String): AgendaKalender {
        return SupabaseManager.client.postgrest["agenda_kalenders"]
            .insert(KalenderInsert("Mijn agenda", "persoonlijk", profielId, "#4F46E5")) { select() }
            .decodeSingle<AgendaKalender>()
    }

    suspend fun afspraken(kalenderIds: List<String>): List<AgendaAfspraak> {
        if (kalenderIds.isEmpty()) return emptyList()
        return SupabaseManager.client.postgrest["agenda_afspraken"]
            .select {
                filter { isIn("kalender_id", kalenderIds) }
                order("start_tijd", Order.ASCENDING)
            }
            .decodeList<AgendaAfspraak>()
    }

    @Serializable
    private data class AfspraakInsert(
        val kalender_id: String,
        val titel: String,
        val beschrijving: String?,
        val start_tijd: String,
        val eind_tijd: String,
        val hele_dag: Boolean,
        val herinnering_minuten: Int?,
        val aangemaakt_door: String,
    )

    suspend fun maakAfspraak(
        kalenderId: String, titel: String, beschrijving: String?,
        startTijd: String, eindTijd: String, heleDag: Boolean,
        herinneringMinuten: Int?, aangemaaktDoor: String,
    ) {
        SupabaseManager.client.postgrest["agenda_afspraken"]
            .insert(AfspraakInsert(kalenderId, titel, beschrijving, startTijd, eindTijd, heleDag, herinneringMinuten, aangemaaktDoor))
    }

    @Serializable
    private data class AfspraakUpdate(
        val kalender_id: String,
        val titel: String,
        val beschrijving: String?,
        val start_tijd: String,
        val eind_tijd: String,
        val hele_dag: Boolean,
        val herinnering_minuten: Int?,
    )

    suspend fun werkAfspraakBij(
        id: String, kalenderId: String, titel: String, beschrijving: String?,
        startTijd: String, eindTijd: String, heleDag: Boolean, herinneringMinuten: Int?,
    ) {
        SupabaseManager.client.postgrest["agenda_afspraken"]
            .update(AfspraakUpdate(kalenderId, titel, beschrijving, startTijd, eindTijd, heleDag, herinneringMinuten)) {
                filter { eq("id", id) }
            }
    }

    suspend fun verwijderAfspraak(id: String) {
        SupabaseManager.client.postgrest["agenda_afspraken"].delete {
            filter { eq("id", id) }
        }
    }
}
